### ISSUE TITLE
DOC: MAST - Limit number of products?

### ISSUE THREAD

**Issue by keflavich** [astropy/astroquery#2507]
I've found that `Observations.get_product_list` takes a long time and returns huge lists for single observations (~300,000 entries that each have many columns).  The documentation presently suggests that one should get that product list then filter it after the fact.

Is there any way to do the filtering before transferring the full metadata?

(this is related to several of the other recent issues about file sizes from MAST, especially https://github.com/astropy/astroquery/issues/2496, but I'm looking for a different workaround)

**Comment by keflavich** [comment]
Even more specifically, I'd like to know if there's a clean way to query for a single file.  i.e., I know that I want the `_asn.json` file for the level-three imaging associated with a specific observation ID.  I can do this, but I have to download and then filter through gigantic files, which makes it difficult to use in a pipeline.

**Comment by bsipocz** [comment]
cc @jaymedina

**Comment by keflavich** [comment]
Followup: the documentation suggests this _is_ possible, but the implemented version doesn't allow it.

https://astroquery.readthedocs.io/en/latest/mast/mast.html#getting-product-lists says "The product fields are documented [here](https://mast.stsci.edu/api/v0/_productsfields.html).", but the `get_product_list` function doesn't pass any of those to MAST.  I don't know if MAST will accept them, but I'll hack away and see...

**Comment by bsipocz** [comment]
I wonder how much of this is the underlying reason for #2505

**Comment by keflavich** [comment]
@bsipocz I bet a lot

ok so in principle, it should work, e.g.:
```python
Observations._parse_result(
Observations._portal_api_connection.service_request_async(Observations._caom_products,
                                                          params={
                                                                  'obs_id': 'jw02221-o001_t001_miri_f2550w',
                                                                  'calib_level': '3',
                                                                  'productSubGroupDescription': 'ASN'})
)
```
but I get 0 results for that.

However, if I do
```python
Observations._parse_result(
Observations._portal_api_connection.service_request_async(Observations._caom_products,
                                                          params={'obsid':'90220705',
                                                                  'obs_id': 'jw02221-o001_t001_miri_f2550w',
                                                                  'calib_level': '3',
                                                                  'productSubGroupDescription': 'ASN'})
)
```
I get 27006 results - independent of what I put in for, e.g., `obs_id`, `calib_level`, ...

So it appears that the MAST `_caom_products` service doesn't respect these keywords and works only on `obsid`?   Unless I'm missing something.

**Comment by keflavich** [comment]
(my examples are with proprietary data, so I'm not sure these should work for others)

**Comment by bsipocz** [comment]
cc @jdavies-st as he worked on addressing https://github.com/astropy/astroquery/issues/2496

**Comment by keflavich** [comment]
Yeah, my complaint very strongly echoes his.

This is what I have to do right now to find a specific `asn` file, which I need to run the pipeline:
```python
filtername='F187N'
output_dir = f'/orange/adamginsburg/jwst/brick/{filtername}/pipeline/'

Observations.cache_location = output_dir
obs_table = Observations.query_criteria(
                                        filters=filtername.lower(),
                                        proposal_id="2221",
                                        proposal_pi="Ginsburg*",
                                        calib_level=3,
                                       )

data_products_by_obs = Observations.get_product_list(obs_table)

products_asn = Observations.filter_products(data_products_by_obs, extension="json")

# there's probably a more efficient way to do this step...
valid_obsids = products_asn['obs_id'][np.char.find(np.unique(products_asn['obs_id']), 'jw02221-o001', ) == 0]

# I've hard-coded the assumption that there is 1 and only 1 match.
match = [x for x in valid_obsids if filtername.lower() in x][0]

asn_mast_data = products_asn[products_asn['obs_id'] == match]

manifest = Observations.download_products(asn_mast_data, download_dir=output_dir)

# MAST creates deep directory structures I don't want
for row in manifest:
    shutil.move(row['Local Path'], output_dir)
```

**Comment by bsipocz** [comment]
> # MAST creates deep directory structures I don't want

Minor side note: Until #2500 is addressed, my workaround for this is to symlink rather than move, otherwise you end up downloading again when you run your code.

**Comment by keflavich** [comment]
> Minor side note: Until https://github.com/astropy/astroquery/issues/2500 is addressed, my workaround for this is to symlink rather than move, otherwise you end up downloading again when you run your code.

Good idea.  (aside: I would have just  🧠 'd your post but gh doesn't do that!)

**Comment by bsipocz** [comment]
another side note: apparently both of us got annoyed by the directory substructure when we've been acting as users, so I'm certain that it was annoying for non-dev users, too. How could we make sure they actually report minor annoyances like this? I would not say that we'll jump on fixing everything, but to rather treat it as feedback as may distill it down to have further improvements. But if no one speaks up then we likely won't know where the annoyances/pain points are.

**Comment by keflavich** [comment]
I don't know.  I have to harangue my students to get them to post bug reports, and they still generally don't do it (students, if you're reading this, post more bug reports!)

**Comment by jdavies-st** [comment]
> I've found that `Observations.get_product_list` takes a long time and returns huge lists for single observations (~300,000 entries that each have many columns). The documentation presently suggests that one should get that product list then filter it after the fact.

The workaround for JWST data that I've found that's pretty quick is to run in a loop.

```python
from astroquery.mast import Observations
from astropy.table import vstack


obs_table = = Observations.query_criteria(proposal_id = '1345', instrument_name = 'NIRCAM')
product_list = [Observations.get_product_list(obs) for obs in obs_table]
products = vstack(product_list)

# Then you can do any further filtering here

```

This keeps the queries to MAST smallish, and prevents the very slow table joins on the server side.  It's pretty trivial to do the `vstack` call on the client side.  Looping through N observations is pretty much always faster than letting the server try to do it.  The timeout usually gets you.

After the fact I found that this is in the MAST documentation for JWST queries here

https://outerspace.stsci.edu/display/MASTDOCS/API+Advanced+Search

but that nugget of advice never made it to the `astroquery.mast` docs.  It probably should, at least until they can improve the performance.  Or make there a way to avoid the 10 times as many AUXILIARY guide star files that are (currently) not very useful for science.

As a side note, I don't know why the association `_asn.json` files are categorized as INFO when they are required to rerun the calibration pipeline, which every scientist will be doing as a matter of course. Things would be a lot easier if they were included with SCIENCE products.  Same with the MSA config files for NIRSpec, which are categorized as AUXILIARY along with all the guide star files, but are also required to run the calibration pipeline.

**Comment by keflavich** [comment]
Thanks for that trick.  I was in the middle of unintentionally downloading ~200,000 guide star observations I didn't need; I'm glad to cut that off now without risk of timing out the product listing.

**Comment by jdavies-st** [comment]
I've been filtering out guide star files (usually an order of magnitude more files than just the science) with the following snippet, following on from the above code:

```python
import numpy as np


# Filter out Guide Star data
mask = np.char.find(products["dataURI"], "_gs-") == -1
products = products[mask]
```

**Comment by bsipocz** [comment]
@keflavich - A lot of improvements have happened to mast in the past week, could you check on this one and see whether there are any action items, whether any more examples mentioned above are missing from the docs, etc?

**Comment by keflavich** [comment]
I'm running the query code right now, but not testing things systematically.  It's still taking a long minute to get the product list.

**Comment by bsipocz** [comment]
Yes, your examples in https://github.com/astropy/astroquery/issues/2507#issuecomment-1235783671 still produces a long list, but what I wonder whether any of the suggestions below that work, and if yes, then those are in the docs yet or not.

**Comment by jdavies-st** [comment]
Doing `Observations.get_product_list` one obs at a time as above is I think good advice for JWST data.  It's not in the `astroquery.mast` docs yet, but should be, given it is the recommended way from the MAST website for `astroquery`, if you can find those docs.

Could there be a special section for a JWST example here?  Or are these docs supposed to be more general?

**Comment by keflavich** [comment]
Special sections are fine. If different missions have to be treated differently, they should be documented as such.  @jdavies-st would you be willing to add these recommendations to the astroquery docs in a new subsection?

**Comment by jdavies-st** [comment]
Yes, no problem.
